### PR TITLE
debug: debug.hideNonDebugHovers

### LINKED ISSUE #48733: Have a way to enable calls to hover providers when debugging

### PR DESCRIPTION
**Pull request by isidorn**
fixes #48733

Introduces a settings `debug.hideNonDebugHovers`. Feedback on the name and description are very welcome.

@alexandrudima as you gave me the pointer I am simply `provideHover` with a 300 ms delay so it does not get spammed. Note that I do not do anything if debug is not active and in a stopped state since then I do not hide the regular hover.

### CODE REVIEW

**Comment by weinand** [comment]
The semantics of `debug.hideNonDebugHovers` violates our "falsy" rule: a missing flag has the same meaning as a flag with value `false`.

What's about `debug.showAllHovers` instead?

**Comment by isidorn** [comment]
@weinand I try to respect that rule for DAP attributes, since the configurationService has nice api to give me the default value I am not so worried from the implementation side.
However for consistency we can change it. Why I was pushing for the negation is that `showAllHovers` as a setting would be a bit strange that the hover is not actually shown due to our implementation. Thus I put the negative spin - not sure if that is smart though

I am open for changing it

**Comment by alexandrudima** [comment]
@isidorn 

How about:
* `debug.suppressLanguageHover`
* `debug.suppressOtherHovers`
* `debug.disableOtherHovers`
* `debug.enableOtherHovers`

I don't know...

**Comment by isidorn** [comment]
@alexandrudima thanks for the suggestions, after discussing with @weinand we landed on 
`debug.enableAllHovers` which by default is `false`
@alexandrudima let me know if the code looks good to you and I will merge it in

File: src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts [hunk 1]
```diff
@@ -340,6 +341,17 @@ export class DebugEditorContribution implements IDebugEditorContribution {
 		return new RunOnceScheduler(() => this.hoverWidget.hide(), HOVER_DELAY);
 	}
 
+	@memoize
+	private get provideNonDebugHoverScheduler(): RunOnceScheduler {
+		return new RunOnceScheduler(() => {
```
Comment (alexandrudima) [inline src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts]: Where is this disposed?

File: src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts [hunk 2]
```diff
@@ -340,6 +341,17 @@ export class DebugEditorContribution implements IDebugEditorContribution {
...
+			const model = this.editor.getModel();
+			const supports = HoverProviderRegistry.ordered(model);
```
Comment (alexandrudima) [inline src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts]: Why not simply call `getHover` (from `getHover.ts`) ?

File: src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts [hunk 3]
```diff
@@ -362,6 +374,10 @@ export class DebugEditorContribution implements IDebugEditorContribution {
 			return;
 		}
 
+		if (!this.configurationService.getValue<IDebugConfiguration>('debug').hideNonDebugHovers) {
+			this.nonDebugHoverPosition = mouseEvent.target.position;
+			this.provideNonDebugHoverScheduler.schedule();
```
Comment (alexandrudima) [inline src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts]: Should we also `.cancel()` this somewhere?

**Review by alexdima — CHANGES_REQUESTED** [review]
Looks good. Some minor feedback.

**Comment by isidorn** [comment]
@alexandrudima @weinand thanks a lot for feedback, I have addressed it. Merging in

**Comment by DanTup** [comment]
What does this setting do? I assumed it would call by normal hover provider (so that I can see the nice documentation on methods when hovering, even in a debug session) but even with the setting set to `true` I'm not seeing them (breakpoint in my hover provider is not hit). I'm a bit confused by the description: "Regular hovers will not be shown even if this setting is true". I'm not sure whether my hover is a "regular hover" and if it is, I'm not sure what this setting is actually enabling.

**Comment by DanTup** [comment]
Actually, it is calling my provider now - but it's still not rendering anything. Are there some rules about when it will render? Should it combine the info from the debugger and the normal hovers? (and what does "Regular hovers will not be shown even if this setting is true" mean?).

**Comment by isidorn** [comment]
Hi @DanTup while a user is debugging all other hovers will not be rendered.
We might change that in the future, but this is the current limiation of our hover implementation that the hovers are unaware of each other and would be rendered on top of one another if we enabled that.

That setting is just such that the extension hover provider still gets called with the position even though the hover would not be rendered. It sounds a bit weird but was a specific use case for one extenstion.

So what you are seeing is designed behavior for now. Thanks

**Comment by DanTup** [comment]
> That setting is just such that the extension hover provider still gets called with the position even though the hover would not be rendered. It sounds a bit weird but was a specific use case for one extenstion.

Aha, gotcha. Thanks for the explanation! :)

**Pull request merged**